fletcher/algorithms/string.py

```diff
@@ -339,32 +339,44 @@ def _slice(offsets, data, start: int, end: int, step: int) -> StringArrayBuilder
     builder = StringArrayBuilder(len(offsets) - 1)
 
     for i in prange(len(offsets) - 1):
+        str_len_bytes = offsets[i + 1] - offsets[i]
+
         char_idx = 0
         byte_idx = 0
 
-        str_len = offsets[i + 1] - offsets[i]
+        while char_idx < start and byte_idx < str_len_bytes:
+            char_idx += 1
+            byte_idx += get_utf8_size(data[offsets[i] + byte_idx])
 
-        # doesn't work
-        # if end < 0:
-        #     end += str_len
-        # if start < 0:
-        #     start += str_len
+        # positive start, end; step > 1
+        if step > 1:
```

Comment:
would probably be a bit better if this were in a different function. I think it would be cleaner to put the logic with step != 1 and step == 1 in the `_slice_handle_chunk` function above.